### PR TITLE
feat(Oauth): Added Oauth Support to Java UAA-Client

### PR DESCRIPTION
**Pull request by jheyl**

### CODE REVIEW

File: src/main/java/io/bosh/client/OAuthCredentialsProvider.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+public class OAuthCredentialsProvider implements Header {
+    private OAuth2AccessToken token;
+    private AccessTokenProviderChain CHAIN = new AccessTokenProviderChain(
+            Arrays.<AccessTokenProvider> asList(new UnsecureClientCredentialsAccessTokenProvider()));
+    private  ClientCredentialsResourceDetails credentials = new ClientCredentialsResourceDetails();
+
+
+    public OAuthCredentialsProvider(String host, String username,
+                                    String password){
+        host="https://" + host + ":8443/oauth/token";
```
Comment (jhiemer) [inline src/main/java/io/bosh/client/OAuthCredentialsProvider.java]: Please add the ability to define the scheme (http vs. https)

Same applies for the Port.

Comment (jhiemer) [reply src/main/java/io/bosh/client/OAuthCredentialsProvider.java]: and URL Builder: 

```java
String protocol = "http";
String host = "example.com";
int port = 4567;
String path = "/foldername/1234";
String auth = null;
String fragment = null;
URI uri = new URI(protocol, auth, host, port, path, query, fragment);
URL url = uri.toURL();
```

File: src/main/java/io/bosh/client/OAuthCredentialsProvider.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    @Override
+    public String getName() {
+        return "Authorization";
+    }
+
+    @Override
+    public String getValue() {
+        return token.getTokenType()+ " "+ token.getValue();
```
Comment (jhiemer) [inline src/main/java/io/bosh/client/OAuthCredentialsProvider.java]: Space between " and +

File: src/main/java/io/bosh/client/SpringDirectorClientBuilder.java [hunk 3]
```diff
@@ -98,10 +103,26 @@ private ClientHttpRequestFactory createRequestFactory(String host, String userna
         SSLConnectionSocketFactory connectionFactory = new SSLConnectionSocketFactory(sslContext,
```
Comment (jhiemer) [inline src/main/java/io/bosh/client/SpringDirectorClientBuilder.java]: Never add this AllowAllHostnameVerifier() as "default". Allow all *can* be enabled upon user request (a flag in application.yml), but should never be the standard configuration

File: src/main/java/io/bosh/client/SpringDirectorClientBuilder.java [hunk 4]
```diff
@@ -98,10 +103,26 @@ private ClientHttpRequestFactory createRequestFactory(String host, String userna
...
 
-        // disabling redirect handling is critical for the way BOSH uses 302's
-        HttpClient httpClient = HttpClientBuilder.create().disableRedirectHandling()
-                .setDefaultCredentialsProvider(credentialsProvider)
-                .setSSLSocketFactory(connectionFactory).build();
+        HttpClient httpClient;
+
+        if(auth.equals(Authentication.BASIC)){
+            CredentialsProvider credentialsProvider = new BasicCredentialsProvider();
+            credentialsProvider.setCredentials(new AuthScope(host, 25555),
```
Comment (jhiemer) [inline src/main/java/io/bosh/client/SpringDirectorClientBuilder.java]: Port should be put into configuration as well.

File: src/main/java/io/bosh/client/SpringDirectorClientBuilder.java (hunk 3 above)
Comment (jheyl) [reply src/main/java/io/bosh/client/SpringDirectorClientBuilder.java]: But no Bosh UAA in a private Network will have a valid ca-cert.

Comment (jhiemer) [reply src/main/java/io/bosh/client/SpringDirectorClientBuilder.java]: Yes, nethertheless this should only be optional. If we all move to lets encrypt in future, those certs will be there.

File: src/main/java/io/bosh/client/SpringDirectorClient.java [hunk 5]
```diff
@@ -63,8 +63,8 @@
         this.jobs = new SpringJobs(restTemplate, root, tasks, deployments);
```
Comment (jhiemer) [inline src/main/java/io/bosh/client/SpringDirectorClient.java]: Why do we prefix Job and VM with Spring? What does it have to do with Spring?

**Pull request merged**